client/services/utils/localSettings.js

```diff
@@ -0,0 +1,143 @@
+const STORAGE_KEY = 'localsettings'
+const DATA_TYPE_KEY = 'dataType'
+const VALUE_KEY = 'value'
+class LocalSettings {
+  constructor ($window) {
+    Object.assign(this, { $window })
+    this.listeners = {}
+    this.STORAGE_KEY = STORAGE_KEY
+    const fromSession = this.get()
+    if (fromSession !== undefined) {
+      this.settingsObject = fromSession
+    } else {
+      this.settingsObject = { }
+    }
+    this.update(this.settingsObject, 'self', () => {})
+  }
+
+  parseSettingString (settingsToUpdate) {
+    const parsedSettings = {}
+    Object.keys(settingsToUpdate).forEach(key => {
+      if (settingsToUpdate[key].hasOwnProperty(DATA_TYPE_KEY) &&
+      settingsToUpdate[key].hasOwnProperty(VALUE_KEY)) {
+        const dataType = settingsToUpdate[key][DATA_TYPE_KEY]
+        // TODO: add the other datatypes
+        // TODO: make it not fail silently
+        switch (dataType.toLowerCase()) {
+          case 'string':
+            parsedSettings[key] = settingsToUpdate[key][VALUE_KEY]
+            break
+          case 'boolean':
+            parsedSettings[key] = (settingsToUpdate[key][VALUE_KEY] === 'true')
+            break
+        }
+      }
+    })
+    return parsedSettings
+  }
+
+  checkSettingValidity (setting) {
+    if (setting.hasOwnProperty('name') && setting.hasOwnProperty('dataType') && setting.hasOwnProperty('value')) {
+      const dataType = setting[DATA_TYPE_KEY]
+      let valid = false
+      if (dataType === 'number' || dataType === 'boolean' || dataType === 'string') {
+        // eslint-disable-next-line valid-typeof
+        valid = (typeof (setting[VALUE_KEY]) === dataType)
+      }
+      return valid
+    }
+    return false
+  }
+
+  addSettings (sourceName, settingsArray) {
+    const settingsObject = this.settingsObject
+    const listeners = this.listeners
+    if (this.settingsObject.hasOwnProperty(sourceName)) {
+      const existingSettings = this.settingsObject[sourceName]
+      settingsArray.forEach(setting => {
+        const settingForUpdate = existingSettings.find(entry => {
+          return entry.name === setting.name
+        })
+        if (settingForUpdate) {
+          const idx = settingsObject[sourceName].indexOf(settingForUpdate)
+          settingsObject[sourceName][idx][VALUE_KEY] = setting[VALUE_KEY]
+        } else {
+          settingsObject[sourceName].push(setting)
+        }
+      })
+    } else {
+      settingsObject[sourceName] = settingsArray
+    }
+    settingsArray.forEach(setting => {
+      if (!listeners.hasOwnProperty(`${sourceName}-${setting.name}`)) {
+        listeners[`${sourceName}-${setting.name}`] = setting.cb
```

Comment:
Remove the callback from the setting before putting it into the `settingsObject`, because saving it is irrelevant.

Reply:
from the `setting` object? not sure I understand why